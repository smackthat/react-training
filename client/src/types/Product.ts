import { TFunction } from 'i18next';
import { translations } from 'locales/translations';

export interface Product {
  id: number;
  title: string;
  price: number;
  description: string;
  category: ProductCategory;
  image: string;
}

export interface ProductsState {
  products: Product[];
  loading: boolean;
  error: boolean;
}

export enum ProductCategory {
  MENS_CLOTHING = `men's clothing`,
  JEWELERY = `jewelery`,
  ELECTRONICS = `electronics`,
  WOMENS_CLOTHING = `women's clothing`,
}

export function ProductCategoryToString(
  category: ProductCategory,
  t: TFunction,
) {
  switch (category) {
    case ProductCategory.ELECTRONICS:
      return t(translations.categories.electronics);
    case ProductCategory.JEWELERY:
      return t(translations.categories.jewelery);
    case ProductCategory.MENS_CLOTHING:
      return t(translations.categories.mensclothing);
    case ProductCategory.WOMENS_CLOTHING:
      return t(translations.categories.womensclothing);
    default:
      return '';
  }
}
