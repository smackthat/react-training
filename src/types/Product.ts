export interface Product {
  id: number;
  title: string;
  price: number;
  description: string;
  category: string;
  image: string;
}

export interface ProductsState {
  products: Product[];
  loading: boolean;
  error: boolean;
}

export enum ProductCategory {
  MENS_CLOTHING = `men's clothing`,
  JEWELERY = `jewelery`,
  ELECTRONICS = `electronics`,
  WOMENS_CLOTHING = `women's clothing`,
}
