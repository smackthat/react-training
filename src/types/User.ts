export interface User {
  id: number;
  userName: string;
}

export interface CartItem {
  productId: number;
  title: string;
  quantity: number;
  sum: number;
}

export interface Cart {
  products: CartItem[];
}

export interface UserState {
  user: User;
  loading: boolean;
  error: UserErrorType[];
  cart?: Cart;
}

export enum UserErrorType {
  RESPONSE_ERROR,
  USER_NOT_FOUND,
  USER_NAME_EMPTY,
  PASSWORD_EMPTY,
}

export interface UserLogin {
  name: string;
  password: string;
}
