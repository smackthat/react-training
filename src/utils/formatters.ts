/** Currency formatter */
export function CurrencyFormatter(): Intl.NumberFormat {
  return new Intl.NumberFormat(undefined, {
    style: 'currency',
    currency: 'EUR',
  });
}
