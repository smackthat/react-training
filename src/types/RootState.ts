import { User } from './User';
// [IMPORT NEW CONTAINERSTATE ABOVE] < Needed for generating containers seamlessly

/* 
  Because the redux-injectors injects your reducers asynchronously somewhere in your code
  You have to declare them here manually
*/
export interface RootState {
  user: User;
  // [INSERT NEW REDUCER KEY ABOVE] < Needed for generating containers seamlessly
}
