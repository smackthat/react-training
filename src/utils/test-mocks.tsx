import { Product, ProductCategory } from 'types/Product';
import { User, UserState } from 'types/User';

export const mockUser: User = {
  id: 2,
  userName: 'TEST',
};

export const mockLoggedInUserState: UserState = {
  user: mockUser,
  loading: false,
  error: null,
  cart: {
    items: [],
  },
};

export const mockProducts: Product[] = [
  {
    id: 1,
    title: 'Shirt',
    description: 'A nice shirt',
    price: 9.99,
    category: ProductCategory.MENS_CLOTHING,
    image: '',
  },
  {
    id: 2,
    title: 'Skirt',
    description: 'A nice skirt',
    price: 19.99,
    category: ProductCategory.WOMENS_CLOTHING,
    image: '',
  },
  {
    id: 3,
    title: 'Phone X',
    description: 'A nice phone',
    price: 999.99,
    category: ProductCategory.ELECTRONICS,
    image: '',
  },
  {
    id: 1,
    title: 'Golden necklace',
    description: 'A nice necklace',
    price: 99.99,
    category: ProductCategory.JEWELERY,
    image: '',
  },
];
