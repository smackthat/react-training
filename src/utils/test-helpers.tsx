import { I18nextProvider } from 'react-i18next';
import React, { ReactElement } from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import i18n from 'locales/i18nForTests';
import { configureAppStore } from 'store/configureStore';
import { RootState } from 'types/RootState';
import { EnhancedStore } from '@reduxjs/toolkit';
import { HelmetProvider } from 'react-helmet-async';

// See: https://testing-library.com/docs/react-testing-library/setup

const defaultState: RootState = {
  productsState: {
    loading: false,
    error: false,
    products: [],
  },
  userState: {
    user: null,
    loading: false,
    error: null,
  },
};

function customRender(
  ui: ReactElement,
  {
    initialState = defaultState,
    store = configureAppStore(initialState),
    ...options
  }: { initialState?: RootState; store?: EnhancedStore } = {},
) {
  return {
    result: render(
      <I18nextProvider i18n={i18n}>
        <HelmetProvider>
          <Provider store={store}>{ui}</Provider>
        </HelmetProvider>
      </I18nextProvider>,
      options,
    ),
    store: store,
  };
}

export * from '@testing-library/react';
export { customRender as render };
