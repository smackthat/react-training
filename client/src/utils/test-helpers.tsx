import { I18nextProvider } from 'react-i18next';
import React, { ReactElement } from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import i18nForTests from 'locales/i18nForTests';
import { configureAppStore } from 'store/configureStore';
import { RootState } from 'types/RootState';
import { EnhancedStore } from '@reduxjs/toolkit';
import { HelmetProvider } from 'react-helmet-async';
import { InjectedReducersType } from './types/injector-typings';
import { userReducer } from 'app/pages/LoginPage/slice';
import { productsReducer } from 'app/pages/HomePage/slice';
import { i18n } from 'locales/i18n';
import { mockProducts } from './test-mocks';

// See: https://testing-library.com/docs/react-testing-library/setup

export const defaultState: RootState = {
  productsState: {
    loading: false,
    error: false,
    products: mockProducts,
  },
  userState: {
    user: null,
    loading: false,
    error: null,
  },
};

const reducers: InjectedReducersType = {
  userState: userReducer,
  productsState: productsReducer,
};

async function customRender(
  ui: ReactElement,
  {
    initialState = defaultState,
    store = configureAppStore(initialState, reducers),
    ...options
  }: { initialState?: RootState; store?: EnhancedStore } = {},
) {
  const t = await i18n;

  return {
    result: render(
      <I18nextProvider i18n={i18nForTests}>
        <HelmetProvider>
          <Provider store={store}>{ui}</Provider>
        </HelmetProvider>
      </I18nextProvider>,
      options,
    ),
    store: store,
    t: t,
  };
}

export * from '@testing-library/react';
export { customRender as render };
