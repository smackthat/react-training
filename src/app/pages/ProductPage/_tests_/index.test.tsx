import * as React from 'react';
import { ProductPage } from '..';
import { MemoryRouter, Route } from 'react-router-dom';
import { translations } from 'locales/translations';
import { i18n } from 'locales/i18n';
import { render } from 'utils/test-helpers';
import { ProductCategory } from 'types/Product';
import { RootState } from 'types/RootState';

const defaultState: RootState = {
  productsState: {
    loading: false,
    error: false,
    products: [
      {
        id: 1,
        title: 'Shirt',
        description: 'A nice shirt',
        price: 9.99,
        category: ProductCategory.MENS_CLOTHING,
        image: '',
      },
    ],
  },
  userState: {
    user: null,
    loading: false,
    error: null,
  },
};
const renderProductPage = (initialState?: RootState) =>
  render(
    <MemoryRouter initialEntries={['product/1']}>
      <Route path="product/:id">
        <ProductPage></ProductPage>
      </Route>
    </MemoryRouter>,
    {
      initialState: initialState || defaultState,
    },
  );

describe('ProductPage', () => {
  it('should match snapshot', () => {
    const productPage = renderProductPage();
    expect(productPage.result).toMatchSnapshot();
  });

  it('should not have Add to cart button when user not logged in', async () => {
    const t = await i18n;

    const productPage = renderProductPage();

    expect(
      productPage.result.queryByText(
        t(translations.product.actions.addToCart) as string,
      ),
    ).not.toBeInTheDocument();
  });

  it('should have Add to cart button when user is logged in', async () => {
    const t = await i18n;

    const productPage = renderProductPage({
      ...defaultState,
      userState: {
        user: {
          id: 2,
          userName: 'TEST',
        },
        loading: false,
        error: null,
      },
    });

    expect(
      productPage.result.queryByText(
        t(translations.product.actions.addToCart) as string,
      ),
    ).toBeInTheDocument();
  });
});
