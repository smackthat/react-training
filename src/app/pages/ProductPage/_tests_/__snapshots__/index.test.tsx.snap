// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductPage should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  width: 75vw;
  margin: 0 auto;
  padding: 0 1.5rem;
  box-sizing: content-box;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-top: 25vh;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  min-height: 320px;
}

.c4 {
  padding: 1em;
}

.c3 {
  max-height: 240px;
  max-width: 300px;
}

.c2 {
  margin-left: 10%;
  width: 70vw;
  padding: 3em;
}

@media (max-width:700px) {
  .c0 {
    width: 90vw;
  }
}

@media (max-width:800px) {
  .c3 {
    max-height: 160px;
    max-width: 130px;
  }
}

<body>
    <div>
      <div
        class="c0"
      >
        <main
          class="c1"
        >
          <div
            class="MuiPaper-root c2 MuiPaper-elevation1 MuiPaper-rounded"
          >
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-8"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
              >
                <img
                  class="c3"
                />
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8 MuiGrid-grid-md-6"
              >
                <div
                  class="c4"
                >
                  <h6
                    class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom"
                  />
                  <h6
                    class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-gutterBottom"
                  >
                    €NaN
                  </h6>
                </div>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8"
              >
                <h6
                  class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-gutterBottom"
                >
                  product.description
                </h6>
                <h6
                  class="MuiTypography-root MuiTypography-subtitle2"
                />
              </div>
            </div>
          </div>
        </main>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="PageWrapper-sc-13dr2o1-0 pBcEs"
    >
      <main
        class="sc-AxjAm gdsAWt"
      >
        <div
          class="MuiPaper-root sc-AxhCb fCehYN MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-8"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
            >
              <img
                class="sc-AxiKw hJSCTo"
              />
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8 MuiGrid-grid-md-6"
            >
              <div
                class="sc-AxirZ bcWlmz"
              >
                <h6
                  class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom"
                />
                <h6
                  class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-gutterBottom"
                >
                  €NaN
                </h6>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-gutterBottom"
              >
                product.description
              </h6>
              <h6
                class="MuiTypography-root MuiTypography-subtitle2"
              />
            </div>
          </div>
        </div>
      </main>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
